source/engine/serialize/ComponentSerialization.cpp

```diff
@@ -0,0 +1,219 @@
+#include "ComponentSerialization.h"
+#include "ncengine/audio/AudioSource.h"
+#include "ncengine/graphics/MeshRenderer.h"
+#include "ncengine/graphics/ParticleEmitter.h"
+#include "ncengine/graphics/PointLight.h"
+#include "ncengine/graphics/ToonRenderer.h"
+#include "ncengine/physics/Collider.h"
+#include "ncengine/physics/ConcaveCollider.h"
+#include "ncengine/physics/PhysicsBody.h"
+#include "ncengine/serialize/SceneSerialization.h"
+#include "physics/ColliderUtility.h"
+
+#include "ncutility/BinarySerialization.h"
+
+namespace nc
+{
+void SerializeAudioSource(std::ostream& stream, const audio::AudioSource& out, const SerializationContext& ctx, void*)
+{
+    serialize::Serialize(stream, ctx.entityMap.at(out.ParentEntity()));
+    serialize::Serialize(stream, out.GetClip());
+    serialize::Serialize(stream, out.GetProperties());
+}
+
+auto DeserializeAudioSource(std::istream& stream, const DeserializationContext& ctx, void*) -> audio::AudioSource
+{
+    auto id = uint32_t{};
+    auto clip = std::string{};
+    auto properties = audio::AudioSourceProperties{};
+    serialize::Deserialize(stream, id);
+    serialize::Deserialize(stream, clip);
+    serialize::Deserialize(stream, properties);
+    return audio::AudioSource{ctx.entityMap.at(id), std::move(clip), std::move(properties)};
+}
+
+void SerializeCollider(std::ostream& stream, const physics::Collider& out, const SerializationContext& ctx, void*)
+{
+    serialize::Serialize(stream, ctx.entityMap.at(out.ParentEntity()));
+    const auto colliderType = out.GetType();
+    serialize::Serialize(stream, colliderType);
+    serialize::Serialize(stream, out.IsTrigger());
+    switch (colliderType)
+    {
+        case physics::ColliderType::Box:
+        {
+            serialize::Serialize(stream, physics::ToBoxProperties(out.GetInfo()));
+            break;
+        }
+        case physics::ColliderType::Sphere:
+        {
+            serialize::Serialize(stream, physics::ToSphereProperties(out.GetInfo()));
+            break;
+        }
+        case physics::ColliderType::Capsule:
+        {
+            serialize::Serialize(stream, physics::ToCapsuleProperties(out.GetInfo()));
+            break;
+        }
+        case physics::ColliderType::Hull:
+        {
+            serialize::Serialize(stream, physics::HullProperties{out.GetAssetPath()});
+            break;
+        }
+    }
+}
+
+auto DeserializeCollider(std::istream& stream, const DeserializationContext& ctx, void*) -> physics::Collider
+{
+    auto id = uint32_t{};
+    auto colliderType = physics::ColliderType{};
+    auto isTrigger = false;
+    serialize::Deserialize(stream, id);
+    serialize::Deserialize(stream, colliderType);
+    serialize::Deserialize(stream, isTrigger);
+    const auto entity = ctx.entityMap.at(id);
+    switch (colliderType)
+    {
+        case physics::ColliderType::Box:
+        {
+            auto properties = physics::BoxProperties{};
+            serialize::Deserialize(stream, properties);
```

Comment:
As an argument for using pointers as out params rather than references - I did really like that the calling code makes it clear via capturing the address that it was an out param.

Reply:
Yeah, I kind of like it too, but ultimately I think static analysis is right. If you're taking a pointer, you should be doing a null check, which results in a lot of tedious "this should never happen, but..." type code. It also looks (at least to me) horrendously out of place for `std` types.